File: examples/search_cascade.py
#!/usr/bin/env python3
"""Demonstrate search access cascade, browser toggle, and multi-engine failover."""

import tarzi


def run_search(engine: str, browser: bool, query: str, limit: int = 3) -> None:
    config = tarzi.Config.load()
    config_str = f"""
[search]
engine = "{engine}"
browser = {str(browser).lower()}
limit = {limit}

[fetcher]
browser = false
"""
    config = tarzi.Config.from_str(config_str)
    search = tarzi.SearchEngine.from_config(config)
    try:
        results = search.search(query, limit)
        print(f"OK engine={engine} browser={browser} results={len(results)}")
        for i, r in enumerate(results, 1):
            print(f"  {i}. {r.title} — {r.url}")
    except Exception as e:
        print(f"ERR engine={engine}: {e}")
    finally:
        search.shutdown()


def main() -> None:
    query = "rust programming language"
    print("=== Bing (browser enabled) ===")
    run_search("bing", True, query)

    print("\n=== DuckDuckGo (browser disabled) ===")
    run_search("duckduckgo", False, query)

    print("\n=== Brave (API if BRAVE_API_KEY set) ===")
    run_search("brave", True, query)

    print("\n=== google_serper (API-only) ===")
    run_search("google_serper", True, query)

    print("\n=== Multi-engine: google_serper,duckduckgo,bing ===")
    run_search("google_serper,duckduckgo,bing", False, query)


if __name__ == "__main__":
    main()

File: examples/search_cascade.rs
//! Demonstrate search access cascade, browser toggle, and multi-engine failover.
//!
//! Run:
//!   cargo run --example search_cascade
//!
//! Optional API keys:
//!   export BRAVE_API_KEY=...
//!   export SERPER_API_KEY=...

use tarzi::{config::Config, search::SearchEngine};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    tracing_subscriber::fmt::init();

    let query = "rust programming language";
    let limit = 3;

    println!("=== Bing (browser enabled) ===");
    run_search("bing", true, query, limit).await;

    println!("\n=== DuckDuckGo (browser disabled) ===");
    run_search("duckduckgo", false, query, limit).await;

    println!("\n=== Brave (API if BRAVE_API_KEY set, else web) ===");
    run_search("brave", true, query, limit).await;

    println!("\n=== google_serper (API-only; skipped without SERPER_API_KEY) ===");
    run_search("google_serper", true, query, limit).await;

    println!("\n=== Multi-engine failover: google_serper,duckduckgo,bing ===");
    run_search("google_serper,duckduckgo,bing", false, query, limit).await;

    Ok(())
}

async fn run_search(engine: &str, browser: bool, query: &str, limit: usize) {
    let mut config = Config::load().unwrap_or_default();
    config.search.engine = engine.to_string();
    config.search.browser = browser;
    config.search.limit = limit;
    config.fetcher.browser = false;

    let mut search = SearchEngine::from_config(&config);
    match search.search(query, limit).await {
        Ok(results) => {
            println!(
                "OK engine={engine} browser={browser} engines={:?} results={}",
                search.engines(),
                results.len()
            );
            for (i, r) in results.iter().enumerate() {
                println!("  {}. {} — {}", i + 1, r.title, r.url);
            }
        }
        Err(e) => println!("ERR engine={engine}: {e}"),
    }
    search.shutdown().await;
}
